lib/spack/spack/solver/asp.py

```diff
@@ -1323,7 +1328,7 @@ def variant_rules(self, pkg):
                         pkg.name, fn.variant_default_value_from_package_py(name, variant.default)
                     )
                 )
-            else:
+            elif not concrete_variant or not concrete_variant.propagate:
```

Comment:
When I tried an example package with `x variant1==v1`, `z` has `variant1` as the default, but `y` has both `v1` and the default for `variant1`.

Reply:
Sorry it sounds like this is before your changes? What I'm saying is:

* I think I understand how the problem occurs without your changes
* I think I see how your changes address that problem
* However, I think your changes introduce a new problem: if you ask for something like `spack spec w ^x variant1==v1`, then Spack would now no longer set a default for `variant1` for `z`.

Or am I missing something about

* somewhere else the default is set
* desired semantics for multi-valued variants don't require a default in this case

?

Reply:
@scheibelp 
This is an example I setup, with my changes, for testing the issue you're talking about. The `foo` variant's default is bar and `a-plus` does get the default value, but so does `b`, along with the propagated value.

` spack -m spec multivalue-variant ^a foo==baz
Input spec
--------------------------------
 -   multivalue-variant
 -       ^a foo==baz

Concretized
--------------------------------
 -   multivalue-variant@2.3%gcc@12.2.0~debug build_system=generic fee=bar foo=bar libs=shared arch=linux-rhel8-zen
 -       ^a@2.0%gcc@12.2.0+bvv~lorem_ipsum build_system=autotools foo=baz foobar=bar arch=linux-rhel8-zen
 -           ^b@1.0%gcc@12.2.0 build_system=generic foo=bar,baz arch=linux-rhel8-zen
 -       ^a-plus@2.0%gcc@12.2.0 build_system=autotools foo=bar arch=linux-rhel8-zen
`

Reply:
Interesting: I see your output has `^b... foo=bar,baz`, does that conflict with the stated desire of this PR? I would have expected the changes here to prefer `^b... foo=baz`

Regarding the output for `a-plus`: that's a good sign, and I could definitely be on the wrong track, but one of the issues with proving this bug exists is that we could get the desired result even if the we don't encode the default value (since the concretizer has to pick a value, it might in fact pick the one we think should be default): if you set a default other than `bar` (or `baz`), does the result still look good?

Reply:
I either misread this or totally ignored that we enter this block for all packages where `not variant.propagate`, which would include `z` (so `z` should set 

The output in https://github.com/spack/spack/pull/39833/files#r1366096054 though highlights that a different problem is occurring (namely, that we aren't propagating the default when changing specs below the root).